sklearn/tests/test_metadata_routing.py

```diff
@@ -0,0 +1,974 @@
+    request = MetadataRequest(owner="nested")
+    request.fit.add_request(param="param", alias=True)
+    router = MetadataRouter(owner="test").add_self(request)
+    assert str(router._self) == str(request)
+    # should be a copy, not the same object
+    assert router._self is not request
+
+    # one can add an estimator as self
+    est = RegressorMetadata().set_fit_request(sample_weight="my_weights")
+    router = MetadataRouter(owner="test").add_self(obj=est)
+    assert str(router._self) == str(est.get_metadata_routing())
+    assert router._self is not est.get_metadata_routing()
+
+    # adding a consumer+router as self should only add the consumer part
+    est = WeightedMetaRegressor(
+        estimator=RegressorMetadata().set_fit_request(sample_weight="nested_weights")
+    )
+    router = MetadataRouter(owner="test").add_self(obj=est)
+    # _get_metadata_request() returns the consumer part of the requests
+    assert str(router._self) == str(est._get_metadata_request())
+    # get_metadata_routing() returns the complete request set, consumer and
+    # router included.
+    assert str(router._self) != str(est.get_metadata_routing())
+    # it should be a copy, not the same object
+    assert router._self is not est._get_metadata_request()
+
+
+def test_metadata_routing_add():
+    # adding one with a string `method_mapping`
+    router = MetadataRouter(owner="test").add(
+        method_mapping="fit",
+        est=RegressorMetadata().set_fit_request(sample_weight="weights"),
+    )
+    assert (
+        str(router)
+        == "{'est': {'mapping': [{'callee': 'fit', 'caller': 'fit'}], 'router': {'fit':"
+        " {'sample_weight': 'weights'}, 'score': {'sample_weight':"
+        " <RequestType.ERROR_IF_PASSED: None>}}}}"
+    )
+
+    # adding one with an instance of MethodMapping
+    router = MetadataRouter(owner="test").add(
+        method_mapping=MethodMapping().add(callee="score", caller="fit"),
+        est=RegressorMetadata().set_score_request(sample_weight=True),
+    )
+    assert (
+        str(router)
+        == "{'est': {'mapping': [{'callee': 'score', 'caller': 'fit'}], 'router':"
+        " {'fit': {'sample_weight': <RequestType.ERROR_IF_PASSED: None>}, 'score':"
+        " {'sample_weight': <RequestType.REQUESTED: True>}}}}"
+    )
+
+
+def test_metadata_routing_get_param_names():
+    router = (
+        MetadataRouter(owner="test")
+        .add_self(
+            WeightedMetaRegressor(estimator=RegressorMetadata()).set_fit_request(
+                sample_weight="self_weights"
+            )
+        )
+        .add(
+            method_mapping="fit",
+            trs=TransformerMetadata().set_fit_request(
+                sample_weight="transform_weights"
+            ),
+        )
+    )
+
+    assert (
+        str(router)
+        == "{'$self': {'fit': {'sample_weight': 'self_weights'}, 'score':"
+        " {'sample_weight': <RequestType.ERROR_IF_PASSED: None>}}, 'trs':"
```

Comment:
Can we give the `RequestType` a nicer `__repr__`?

Once there is a nicer `__repr__` for `RequestType`, the metadata routing almost looks like a json dict. I know I pushed back on allowing for a dictionary return for `get_metadata_requests`, but I suspect it will be useful third party estimators.  **This point is not a blocker for this PR.**

Reply:
In the same way that we should have a nice representation for `estimator.get_params()`, we should have a nice representation.
I am thinking to propose a graphical one at some points.